### PR TITLE
feat: Initial Jenkins shared library

### PR DESCRIPTION
**Pull request by Wayonb**
The Jenkins shared library is to allow sharing of code between multiple jobs.  This contains the default pipeline code for ci with some generic steps.  The library will be expanded in the future to support docker builds and plus infrastructure rollouts.

### CODE REVIEW

File: jenkins/shared-library/vars/defaultCiPipeline.groovy [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            RELEASE_BRANCH = 'main'
+            BUILD_SETUP_SCRIPT_FILEPATH = 'scripts/ci/setup_build.sh'
+            BUILD_SCRIPT_FILEPATH = 'scripts/ci/build.sh'
+            TEST_SETUP_SCRIPT_FILEPATH = 'scripts/ci/setup_tests.sh'
+            TEST_SCRIPT_FILEPATH = 'scripts/ci/test.sh'
+            LINTER_SCRIPT_FILEPATH = 'scripts/ci/linter.sh'
+        }
+
+        stages {
+            stage('CI pipeline') {
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/defaultCiPipeline.groovy]: do you envision this replacing most of catapult-client jenkins files?

File: jenkins/shared-library/vars/fileHelper.groovy [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    return Paths.get("${WORKSPACE_TMP}", filepath)
+}
+
+String copyToTempFile(String data, String destinationFilepath) {
+    return copyToLocalFile(data, getTempPathInWorkspace(destinationFilepath))
+}
+
+String copyToLocalFile(String data, String destinationFilepath) {
+    String parentPath = Paths.get(destinationFilepath).getParent()
+    dir (parentPath) {
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/fileHelper.groovy]: why do you need to execute in parentPath? aren't you providing full filepath to writeFile?

File: jenkins/shared-library/vars/logger.groovy [hunk 3]
```diff
@@ -0,0 +1,11 @@
+def logError(String message) {
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/logger.groovy:1]: can you change the def returns to something else? void/Object/etc ?

File: jenkins/shared-library/resources/publish_artifacts/scripts/node-functions.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  VERSION="$(npm run version --silent)"
+  validate_env_variable "VERSION"
+  echo -e "$VERSION"
+}
+
+node_publish_alpha() {
+  VERSION="$(node_load_version)"
+  validate_env_variable "VERSION"
+  validate_env_variable "NPM_TOKEN"
+  NEW_VERSION="$VERSION-$(date +%Y%m%d%H%M)"
```
Comment (Jaguar0625) [inline jenkins/shared-library/resources/publish_artifacts/scripts/node-functions.sh]: we want to drop date, don't we?

File: jenkins/shared-library/resources/publish_artifacts/scripts/docker-functions.sh [hunk 5]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
```
Comment (Jaguar0625) [inline jenkins/shared-library/resources/publish_artifacts/scripts/docker-functions.sh:1]: do we need docker-functions separate from docker-publish?

File: jenkins/shared-library/resources/publish_artifacts/scripts/jenkins-functions.sh [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  CHECKOUT_BRANCH="$1"
+  validate_env_variable "GIT_URL"
+  validate_env_variable "CHECKOUT_BRANCH"
+  validate_env_variable "GITHUB_TOKEN"
+  validate_env_variable "REMOTE_NAME"
+
+  git remote rm $REMOTE_NAME
+  echo "Setting remote url $GIT_URL"
+  git remote add $REMOTE_NAME "$GIT_URL" >/dev/null 2>&1
+  echo "Checking out $CHECKOUT_BRANCH as travis leaves the head detached."
```
Comment (Jaguar0625) [inline jenkins/shared-library/resources/publish_artifacts/scripts/jenkins-functions.sh]: update travis => CI ?

**Review by Jaguar0625 — CHANGES_REQUESTED** [review]
also would probably go through the 50-100 info-level linter warnings as they seem to be mostly good suggestions

File: jenkins/shared-library/README.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+	dockerImageName = 'symbolplatform/symbol-server-private'
+	gitHubId = 'Symbol-Github-app'
+}
+````
+
+``platform`` - is an array of OS the package needs to build on.
+
+``dockerImageName`` - The name of the docker image to publish
+
+``npm`` - set to true to publish npm packages
```
[the remark this answers is not in the document]
Comment (Wayonb) [reply jenkins/shared-library/README.md:21]: make sense.  Will add it.

File: jenkins/shared-library/vars/defaultCiPipeline.groovy (hunk 1 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/defaultCiPipeline.groovy]: maybe some of them

File: jenkins/shared-library/resources/publish_artifacts/scripts/node-functions.sh (hunk 4 above)
Comment (Wayonb) [reply jenkins/shared-library/resources/publish_artifacts/scripts/node-functions.sh]: yea, Iet me see how to add the hash

File: jenkins/shared-library/resources/publish_artifacts/scripts/docker-functions.sh (hunk 5 above)
Comment (Wayonb) [reply jenkins/shared-library/resources/publish_artifacts/scripts/docker-functions.sh:1]: no, but I will leave it until I rewrite the all the shell scripts to remove them all.

File: jenkins/shared-library/resources/publish_artifacts/scripts/docker-functions.sh [hunk 8]
```diff
@@ -36,6 +34,8 @@ docker_build() {
 			docker push "${DOCKER_IMAGE_NAME}:latest"
 			docker push "${DOCKER_IMAGE_VERSION_NAME}"
 		fi
+    DIGESTS=$(docker inspect --format='{{index .RepoDigests 0}}' "${DOCKER_IMAGE_VERSION_NAME}")
+    echo "The DIGESTS is ${DIGESTS}"
```
Comment (Jaguar0625) [inline jenkins/shared-library/resources/publish_artifacts/scripts/docker-functions.sh:38]: spaces => tabs

File: jenkins/shared-library/vars/defaultCiPipeline.groovy [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+						when {
+							anyOf {
+								branch env.DEV_BRANCH
+								allOf {
+									expression {
+										return env.SHOULD_PUBLISH_IMAGE.toBoolean()
+									}
+									not {
+										branch env.RELEASE_BRANCH
+									}
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/defaultCiPipeline.groovy:131]: why did you need this? aren't you checking it is DEV_BRANCH above?

File: jenkins/shared-library/resources/artifacts/scripts/docker-functions.sh [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+		fi
+
+		if [ "$OPERATION" = "release" ]; then
+			docker tag "${DOCKER_IMAGE_VERSION_NAME}" "${DOCKER_IMAGE_NAME}:latest"
+			echo "Docker pushing release"
+			docker push "${DOCKER_IMAGE_NAME}:latest"
+			docker push "${DOCKER_IMAGE_VERSION_NAME}"
+		fi
+    DIGESTS=$(docker inspect --format='{{index .RepoDigests 0}}' "${DOCKER_IMAGE_VERSION_NAME}")
+    echo "The DIGESTS is ${DIGESTS}"
```
Comment (Jaguar0625) [inline jenkins/shared-library/resources/artifacts/scripts/docker-functions.sh:38]: fix indentation (spaces => tabs)

File: jenkins/shared-library/vars/publish.groovy [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+void copyFileListToTemp(String[] files) {
+	files.each { filepath ->
+		fileHelper.copyToTempFile(getResource(Paths.get('artifacts').resolve(filepath).toString()), filepath)
+	}
+}
+
+void copyWithJenkinsFunctionFileToTemp(String[] files) {
+	copyFileListToTemp(files.plus(['scripts/jenkins-functions.sh']) as String[])
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/publish.groovy:20]: (files + ['scripts/jenkins-functions.sh'])

File: jenkins/shared-library/vars/defaultCiPipeline.groovy [hunk 12]
```diff
@@ -0,0 +1,182 @@
+/* groovylint-disable NestedBlockDepth */
+import java.nio.file.Paths
+
+void call(Closure body) {
+	Map params = [:]
+	body.resolveStrategy = Closure.DELEGATE_FIRST
+	body.delegate = params
+	body()
+
+	String packageRootPath = getJenkinsfilePath()
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/defaultCiPipeline.groovy:10]: should we suppress `UnnecessaryGetter` lint ?

File: jenkins/shared-library/vars/defaultCiPipeline.groovy (hunk 9 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/defaultCiPipeline.groovy:131]: The ``anyOf`` is OR condition.  Its expands to ``( branch == dev)  or (should_publish_image && not main)``.  
While writing this makes me think of another scenario, should be able to release a RC build from main.  I will add another option which is cleaner

File: jenkins/shared-library/resources/artifacts/scripts/docker-functions.sh (hunk 10 above)
Comment (Wayonb) [reply jenkins/shared-library/resources/artifacts/scripts/docker-functions.sh:38]: ok.

File: jenkins/shared-library/vars/publish.groovy (hunk 11 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/publish.groovy:20]: yea, the linter was complaining.  I am assuming its due to ``['scripts/jenkins-functions.sh']`` is an object array instead of String[]

File: jenkins/shared-library/vars/defaultCiPipeline.groovy (hunk 12 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/defaultCiPipeline.groovy:10]: ok

**Review by Jaguar0625 — COMMENTED** [review]
made a few tweaks to the linter on top

File: linters/groovy/.groovylintrc.json [hunk 13]
```diff
@@ -8,6 +8,8 @@
 
     "size.ParameterCount": { "maxParameters": 9 },
 
-    "dry.DuplicateStringLiteral": { "ignoreStrings": ",/,Jenkinsfile" }
+    "dry.DuplicateStringLiteral": { "ignoreStrings": ",/,\n,Jenkinsfile" },
+
+    "unnecessary.UnnecessaryGetter": { "checkIsMethods": false, "ignoreMethodNames": "getJenkinsfilePath" }
```
Comment (Jaguar0625) [inline linters/groovy/.groovylintrc.json:13]: probably would rename `getJenkinsfilePath` to something like `findJenkinsfilePath`

**Review by Jaguar0625 — APPROVED** [review]
